R/check-control.R

```diff
@@ -0,0 +1,305 @@
+                            expr_test = NULL,                          
+                            not_found_msg = NULL) {
+  cond_test <- substitute(cond_test)
+  expr_test <- substitute(expr_test)
+  test_loop(index = index, 
+            cond_test = cond_test,
+            expr_test = expr_test,
+            not_found_msg = not_found_msg,
+            fun = check_while)
+}
+
+#' @rdname test_control
+#' @export
+test_for_loop <- function(index = 1, 
+                          cond_test = NULL,
+                          expr_test = NULL,
+                          not_found_msg = NULL) {
+  cond_test <- substitute(cond_test)
+  expr_test <- substitute(expr_test)
+  test_loop(index = index, 
+            cond_test = cond_test,
+            expr_test = expr_test,
+            not_found_msg = not_found_msg,
+            fun = check_for)
+}
+
+test_loop <- function(index, cond_test, expr_test, not_found_msg, fun) {
+  old_state <- ex()
+  test_env <- old_state$get("test_env")
+  on.exit(tw$set(state = old_state))
+  
+  testloop <- old_state %>% fun(index = index, not_found_msg = not_found_msg, append = is.null(not_found_msg))
+  
+  if (!is.null(cond_test)) {
+    cond_state <- testloop %>% check_cond()
+    tw$set(state = cond_state)
+    eval(cond_test, envir = test_env)
+  }
+  
+  if (!is.null(expr_test)) {
+    body_state <- testloop %>% check_body()
+    tw$set(state = body_state)
+    eval(expr_test, envir = test_env)
+  }
+}
+
+
+
+
+#' @rdname test_control
+#' @export
+check_if_else <- function(state, index = 1, not_found_msg = NULL, append = TRUE) {
+  check_control(state, index, not_found_msg, append = append, type = "if")
+}
+
+#' @rdname test_control
+#' @export
+check_while <- function(state, index = 1, not_found_msg = NULL, append = TRUE) {
+  check_control(state, index, not_found_msg, append = append, type = "while")
+}
+
+#' @rdname test_control
+#' @export
+check_for <- function(state, index = 1, not_found_msg = NULL, append = TRUE) {
+  check_control(state, index, not_found_msg, append = append, type = "for")
+}
+
+
+check_control <- function(state, index, not_found_msg, append, type = c("if", "while", "for")) {
+  type <- match.arg(type)
+  student_pd <- state$get("student_pd")
+  solution_pd <- state$get("solution_pd")
+  student_code <- state$get("student_code")
+  solution_code <- state$get("solution_code")
+  
+  control_state <- ControlState$new(state)
+  control_state$add_details(type = type,
+                            case = "defined",
+                            index = index,
+                            message = not_found_msg,
+                            append = append,
+                            pd = NULL)
+
+  extract_fun <- switch(type,
+                        `if` = extract_if,
+                        `for` = extract_for,
+                        `while` = extract_while)
+  student_structs <- extract_fun(student_pd)
+  solution_structs <- extract_fun(solution_pd)
```

Comment:
I'm wondering whether you could make `extract()` into a S3 function which takes a state as a first argument and enriches this with extracted statements. 
E.g.: 
```
extract.ControlState <- function(state, type, ...) {
  statement_key <- paste0(type,"_statements")
  control_statements <- state$get(statement_key)
  if (is.null(control_statement)) {
    return(control_statements)
  } else {
    ... # Extraction logic depending on type
    state$set(statement_key = control_statements)
    return(control_statements)
  }
}
```
Then you could add an extractor per State, seems semantically logical (now there's a `find_function_calls` somewhere, which is actually the function call extractor I think?). Also it's reusable if you want to check multiple things on the same state.